### ISSUE TITLE
Add Support for AutoClusterFailover

### ISSUE THREAD

**Issue by scarlier** [spring-projects/spring-pulsar#375]
**Problem**
I want to have my pulsar client automatically fail over to another cluster when there are connectivity issues.

**Solution**
Pulsar supports AutoClusterFailover since version 2.10. Implementing this feature would solve the problem.
https://pulsar.apache.org/docs/2.10.x/client-libraries-java/#how-to-use-cluster-level-failover

**Comment by onobc** [comment]
HI @scarlier ,
That does sound like a nice feature to support. We will take a look.

**Comment by onobc** [comment]
A quick mental-mapping leads me to a possible solution that would look as follows in the `application.yml`. 

```yaml
spring:              
  pulsar:
    client:
      service-url: "pulsar+ssl://localhost:6650"
      auth-plugin-class-name: TokenAuth
      authentication:
        token: "foo6650"

      failover:
        primary: "pulsar+ssl://localhost:6650"
        failover-delay: 30s
        switch-back-delay: 60s
        check-interval: 1s
        backup-clusters:
          
          - service-url: "pulsar+ssl://localhost:6651"
            auth-plugin-class-name: TokenAuth
            auth-params: "{\"token\":\"foo6651\"}"
            tlsTrustCertsFilePath: "..."
            tlsTrustStorePath: "..."
            tlsTrustStorePassword: "..."

          - service-url: "pulsar+ssl://localhost:6652"
            auth-plugin-class-name: SomeOtherAuth
            auth-params: "{\"shiny\":\"foo6652\"}"
            tlsTrustCertsFilePath: "..."
            tlsTrustStorePath: "..."
            tlsTrustStorePassword: "..."
```

It would be fairly straightforward to implement. We may not have the cycles to get to it in the next few weeks though. If you need it sooner @scarlier and are willing to contribute, I would be glad to point you in the right direction w/ the codebase etc.. 

Thanks for requesting this feature!

**Comment by scarlier** [comment]
Thanks @onobc , currently also full on work. let's see in coming weeks who has a bit of free time :)

**Comment by swamymavuri** [comment]
Hi @onobc ,
Can you please assign this issue to me.
Thanks

**Comment by onobc** [comment]
Hi @swamymavuri ,

I am glad you are interested in working on this and have assigned it to you. 

So, things have changed a bit in the project since the ticket was created. Namely, the auto-configuration has moved into the Spring Boot project (generally available in [3.2.0](https://github.com/spring-projects/spring-boot/milestone/308)) . As such, these code changes will now be in Spring Boot. 

Here is what I think needs to be done (at a high level).

* Add new `Failover` properties under [Client config props](https://github.com/spring-projects/spring-boot/blob/ac4b67465a9644c480d5dad4045a38bad4a4e8e0/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/PulsarProperties.java#L104). Something like the following:
```yaml
spring:
  pulsar:
    client:
      service-url: "pulsar+ssl://localhost:6650"
      authentication:
        plugin-class-name: TokenAuth
        param:
          token: "foo6650"
      failover:
        failover-delay: 30s
        switch-back-delay: 60s
        check-interval: 1s
        backup-clusters:
          - service-url: "pulsar+ssl://localhost:6651"
            authentication:
              plugin-class-name: TokenAuth
              param:
                token: "foo6651"
          - service-url: "pulsar+ssl://localhost:6652"
            authentication:
              plugin-class-name: TokenAuth
              param:
                token: "foo6652"
```

* Consume new properties in the [Client customizer](https://github.com/spring-projects/spring-boot/blob/ac4b67465a9644c480d5dad4045a38bad4a4e8e0/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/PulsarPropertiesMapper.java#L59). Probably something like the following:
```java
void customizeClientBuilder(ClientBuilder clientBuilder, PulsarConnectionDetails connectionDetails) {
	PulsarProperties.Client properties = this.properties.getClient();
	PropertyMapper map = PropertyMapper.get().alwaysApplyingWhenNonNull();
	map.from(connectionDetails::getBrokerUrl).to(clientBuilder::serviceUrl);
	map.from(properties::getConnectionTimeout).to(timeoutProperty(clientBuilder::connectionTimeout));
	map.from(properties::getOperationTimeout).to(timeoutProperty(clientBuilder::operationTimeout));
	map.from(properties::getLookupTimeout).to(timeoutProperty(clientBuilder::lookupTimeout));
	customizeAuthentication(clientBuilder::authentication, properties.getAuthentication());
	customizeServiceUrlProviderBuilder(clientBuilder::serviceUrlProvider, properties);
}

private void customizeServiceUrlProviderBuilder(Consumer<ServiceUrlProvider> serviceUrlProviderConsumer, 
   PulsarProperties.Client properties) {
        // TODO build the authentications and add to map of url->auth below
	Map<String, Authentication> secondaryAuths = new HashMap<>();
	AutoClusterFailoverBuilder failoverBuilder = AutoClusterFailover.builder()
			.primary(properties.getServiceUrl())
			.failoverDelay(30, TimeUnit.SECONDS) // TODO get these values from newly added props
			.switchBackDelay(60, TimeUnit.SECONDS)
			.checkInterval(1000, TimeUnit.MILLISECONDS)
			.secondary(List.of("<<get these from 'spring.pulsar.client.failover.backup-clusters.service-url>>"))
			.secondaryAuthentication(secondaryAuths);
	serviceUrlProviderConsumer.accept(failoverBuilder.build());
}


```

* Add tests for new [Client auto-configuration](https://github.com/spring-projects/spring-boot/blob/fd5722106f5892cf55b8fa49600f95b5b931d5ea/spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/pulsar/PulsarConfigurationTests.java#L81). 

Here are the [latest docs](https://pulsar.apache.org/docs/3.1.x/client-libraries-cluster-level-failover/#automatic-failover) on the automatic failover.

Let me know if you have any questions etc.. 

Thanks

**Comment by swamymavuri** [comment]
Thanks @onobc for pointing the code snippets and the documentation.
I will look into it and get back if i have any questions.

**Comment by swamymavuri** [comment]
Hi @onobc ,


PR for autocluster failover.
https://github.com/spring-projects/spring-boot/pull/38559


https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ClientBuilderImpl.java#L57
This gives the conflict with the PulsarClient, either we need to set serviceUrlProvider or serviceUrl, but not both
This proposed change will effectively resolve the conflict by allowing the setting of either serviceUrl or serviceUrlProvider based on the presence of backup clusters, ensuring proper failover configurations without causing conflicts within the Pulsar client setup.

**Comment by onobc** [comment]
Nice work @swamymavuri,

Your Spring Boot contribution was merged and will be available in Spring Boot 3.3.0-M1 which will be released next Thursday 2024-01-18.

Closing via https://github.com/spring-projects/spring-boot/commit/a2b10e06ea20d62f031be2a93bd9aad142c7f4a8

**Comment by swamymavuri** [comment]
Thanks @onobc for helping me in clarifying the doubts
